chat-plugins/trivia.js

```diff
@@ -0,0 +1,1225 @@
+/*
+ * Trivia plugin
+ * Written by Morfent
+ */
+
+'use strict';
+
+const fs = require('fs');
+
+// Hack to allow unit tests to run.
+const RoomGame = require('../room-game.js').RoomGame;
+const RoomGamePlayer = require('../room-game.js').RoomGamePlayer;
+
+const CATEGORIES = {
+	ae: 'Arts and Entertainment',
+	pokemon: 'Pok\u00E9mon',
+	sg: 'Science and Geography',
+	sh: 'Society and History',
+};
+
+const MODES = {
+	first: 'First',
+	number: 'Number',
+	timer: 'Timer',
+};
+
+// NOTE: trivia code depends on this object's values being divisible by 5.
+const SCORE_CAPS = {
+	short: 20,
+	medium: 35,
+	long: 50,
+};
+
+const SIGNUP_PHASE = 'signups';
+const QUESTION_PHASE = 'question';
+const INTERMISSION_PHASE = 'intermission';
+
+const MINIMUM_PLAYERS = 3;
+
+const START_TIMEOUT = 30 * 1000;
+const QUESTION_INTERVAL = 15 * 1000;
+const INTERMISSION_INTERVAL = 30 * 1000;
+
+const MAX_QUESTION_LENGTH = 252;
+const MAX_ANSWER_LENGTH = 32;
+
+// FIXME: move trivia database code to a separate file once relevant.
+let triviaData = {};
+try {
+	triviaData = require('../config/chat-plugins/triviadata.json');
+} catch (e) {} // file doesn't exist or contains invalid JSON
+if (!triviaData || typeof triviaData !== 'object') triviaData = {};
+if (!triviaData || typeof triviaData.leaderboard !== 'object') triviaData.leaderboard = {};
+if (!Array.isArray(triviaData.ladder)) triviaData.ladder = [];
+if (!Array.isArray(triviaData.questions)) triviaData.questions = [];
+if (!Array.isArray(triviaData.submissions)) triviaData.submissions = [];
+
+const writeTriviaData = (() => {
+	let writing = false;
+	let writePending = false; // whether or not a new write is pending
+
+	let finishWriting = function () {
+		writing = false;
+		if (writePending) {
+			writePending = false;
+			writeTriviaData();
+		}
+	};
+	return function () {
+		if (writing) {
+			writePending = true;
+			return;
+		}
+		writing = true;
+		let data = JSON.stringify(triviaData, null, 2);
+		fs.writeFile('config/chat-plugins/triviadata.json.0', data, () => {
+			// rename is atomic on POSIX, but will throw an error on Windows
+			fs.rename('config/chat-plugins/triviadata.json.0', 'config/chat-plugins/triviadata.json', err => {
+				if (err) {
+					// This should only happen on Windows
+					fs.writeFile('config/chat-plugins/triviadata.json', data, finishWriting);
+					return;
+				}
+				finishWriting();
+			});
+		});
+	};
+})();
+
+// Binary search for the index at which to splice in new questions in a category,
+// or the index at which to slice up to for a category's questions
+function findEndOfCategory(category, inSubmissions) {
+	let questions = inSubmissions ? triviaData.submissions : triviaData.questions;
+	let left = 0;
+	let right = questions.length - 1;
```

Comment:
[a, b) means the range from a (inclusive) to b (exclusive), while [a, b] is the range from a to b (both inclusive).

In this case, [left, right) means starting out with `left = 0` and `right = questions.length` (and modifying the rest of the code to account for that)